router/Route.js

```diff
@@ -341,11 +341,30 @@ export default class Route {
 	_substituteOptionalParamInPath(path, paramName, paramValue) {
 		return path.replace(
 			new RegExp(`(^|\/):\\\?${paramName}([\/\?]|$)`),
-			paramValue ? '$1' + paramValue +  '$2' : ''
+			paramValue ? '$1' + this._encodePathParamValue(paramValue) +  '$2' : ''
 		);
 	}
 
 	/**
+	 * Encode special characters of path parameter value.
+	 * 
+	 * @param {string|number} paramValue Path parameter value
+	 * @return {string} Encoded path parameter value.
+	 */
+	_encodePathParamValue(paramValue) {
+		// Percen sing (%) has to be first!
+		const charsToEncode = ['%', '/', '?', '#', ':'];
+
+		let newParamValue = String(paramValue);
+
+		charsToEncode.forEach(char => {
+			newParamValue = newParamValue.replace(char, encodeURIComponent(char));
```

Comment:
1. This is unnecessary repeated computation. Please add a module-scoped constant `PARAMETER_CHARS_TABLE`:

```javascript
PARAMETER_CHARS_TABLE = new Map(PARAMETER_CHARS_TO_ENCODE.map(char => [char, encodeURIComponent(char)]))
```

2. For-of is preferable in ima's code base to .forEach, please change this.